prov/efa/src/efa_device.c

```diff
@@ -178,34 +178,46 @@ int efa_device_list_initialize(void)
 
 	initialized = true;
 
-	ibv_device_list = ibv_get_device_list(&g_device_cnt);
+	ibv_device_list = ibv_get_device_list(&total_device_cnt);
 	if (ibv_device_list == NULL)
 		return -FI_ENOMEM;
 
-	if (g_device_cnt <= 0) {
+	if (total_device_cnt <= 0) {
 		ibv_free_device_list(ibv_device_list);
 		return -FI_ENODEV;
 	}
 
-	g_device_list = calloc(g_device_cnt, sizeof(struct efa_device));
+	g_device_list = calloc(total_device_cnt, sizeof(struct efa_device));
 	if (!g_device_list) {
 		ret = -FI_ENOMEM;
 		goto err_free;
 	}
 
-	for (device_idx = 0; device_idx < g_device_cnt; device_idx++) {
-		err = efa_device_construct(&g_device_list[device_idx], ibv_device_list[device_idx]);
+	/*
+	 * Read the environment variable FI_EFA_IFACE and only initialize
+	 * devices that match the requested devices
+	 */
+	for (device_idx = 0; device_idx < total_device_cnt; device_idx++) {
+		if (!efa_env_allows_nic(ibv_device_list[device_idx]->name))
+			continue;
+
+		err = efa_device_construct(&g_device_list[g_device_cnt], ibv_device_list[device_idx]);
 		if (err) {
 			ret = err;
 			goto err_free;
 		}
+		g_device_cnt++;
```

Comment:
Are we sure this is the right functional change?  We are changing g_device_cnt to be the max number of NIC's on the platform to the max number of NIC's libfabric can use.  Have you looked at all the other places g_device_cnt is used and confirmed we want this behavior?

Reply:
Thank you for forcing me to double check. I missed the usage in `efa_av.c` which requires all EFA devices. I've updated the PR.